### ISSUE TITLE
Simblee board files for Arduino - bad links, missing from doc

### ISSUE THREAD

**Issue by wjcroft** [OpenBCI/OpenBCI_Ganglion_Library#17]
Shirley quickly replied to this thread with a link to Google Drive with the zipped Arduino board files. But should this be in the repo, or somewhere official?

https://openbci.com/forum/index.php?p=/discussion/2028/simblee-board-files-for-arduino-missing

Dead link for Simblee Arduino download here,

http://docs.openbci.com/Hardware/09-Ganglion_Programming_Tutorial

Shirley's upload, (thanks!):

https://drive.google.com/open?id=1aZijlZ081pie5su9li1vdSes1nh_IMtv

I notice the zip is about 400 megabytes and expands to over 1.5 GB. Gulp. Is that including more than what is needed? Seems rather heavyweight compared to other Arduino IDE board files I've seen, which are only in the 10's of megabytes in size. This zip includes both PIC32 gcc, and Arm gcc plus all libs.

**Comment by wjcroft** [comment]
All the files on simblee.com are now gone. However a web search turned up this copy of the (slightly edited) original Simblee Arduino Board Manager master file, 

http://www.slinq.eu/package_simblee_index.json

from

http://slinq.eu/?page_id=11

Seems like this could be customized for our case, just removing unneeded sections. It points to all the required compilers and toolchains. That are downloaded from Arduino.cc. Or... one of the lab computers might have a copy of the original board manager json.

Another example of a customized Simblee Arduino board manager json,

https://github.com/OpenHAK/Simblee/blob/master/package_openhak_index.json

**Comment by wjcroft** [comment]
A copy of the Simblee Quickstart pdf, which you should save in our repo, not leave on Archive (as it could disappear).

https://web.archive.org/web/20171113001948/https://www.simblee.com/Simblee_Quickstart_Guide_v1.0.pdf

Assuming you will advise anyone using it to also get the json from our repo.

**Comment by shirleyzhang867** [comment]
Hi William,

The zipped file is large because it contains the whole file pathway to get to package_simblee166_index. Before Simblee support was discontinued, you would go to Arduino preferences --> additional board managers --> and insert the URL www.simblee.com/package_simblee166_index.json., and the package would be retrieved. Now in order to program the ganglion, you need to manually install the simblee files in the right folder. The zipped folder also includes the variant.h folder mentioned in the [Ganglion programming tutorial.](http://docs.openbci.com/Hardware/09-Ganglion_Programming_Tutorial#ganglion-programming-tutorial-building-from-source-modify-the-simblee-board-files

**Comment by shirleyzhang867** [comment]
We ran into issues when we tried to set-up (from scratch) a desktop with the files, compilers, and tools needed to program Ganglions. My Macbook was the only computer at our office (and possibly the world) with the Simblee files already installed and able to program Ganglions, so we copied the entire folder "Arduino15" and transferred it to Conor's Macbook via USB drive. And now his Macbook is set-up. I agree the documentation should be updated with a download link to this zip file, but I'd like to look into this a little more before I drop the link into the ganglion programming tutorial.

**Comment by wjcroft** [comment]
Shirley, thanks.

Here is the reason I mentioned the possibility of trying to allow again the ability to use an Arduino board manager setup using our own customized json file, in this case located on our own server.

The unpacked zip file you posted on Google Drive, is 1.6 gigabytes in total size. That is WAY larger than it needs to be for Simblee development. Here is why. The zip contains:

* 1 gigabyte of Chipkit / PIC32 toolchains, compiler, library, etc. Simblee is not this architecture, Simblee is Arm based. Hence this is 1 gigabyte of stuff not relevant to Simblee programmers. Folder in the zip: packages/chipKIT

* 350 megabytes of Simblee toolchains, this is relevant. Folder: packages/Simblee

* 230 megabytes, folder: staging/packages. Which contains the original zip files for the above packages root folders. Contents:

- chipKIT_USB_drivers_otherOSes_v2.zip
- chipkit-core-16778048.master-v1.4.3.zip
- gcc-arm-none-eabi-4.8.3-2014q1-mac.tar.gz
- pic32-tools-143-3-Darwin-image.tar.gz
- pic32prog-macosx-2.1.24.tar.gz
- simblee_253.tar.gz

So in this list, only the gcc-arm and simblee gz files are relevant. And Simblee compilation with this zip is only supported on Mac, not Windows. (gcc-arm is mac based.)

Anyway, my suggestion before finalizing the zip available to customers, is to trim off the extraneous files (over a gigabyte).

----

Additionally, given the json files in the root of this zip folder, it looks very possible to reconstruct a working Arduino IDE board manager json, based on, the original one from Simblee,

package_simblee166_index.json

That file actually references simblee_253.tar.gz, (on simblee.com) which I've uploaded to our own iPage server, at:

http://openbci.com/arduino/simblee_253.tar.gz

Which could be referenced by an edited version of the json, located here for access by the Arduino IDE board manager:

http://openbci.com/arduino/package_simblee166_index.json

Basically what we would do is edit the json to eliminate all but the latest version Simblee pointers, then point that at our own tar.gz mentioned above.

The advantage of this approach is that it preserves the rest of the json file, which pulls in the toolchains for the various different operating systems. That part of the json would be left alone.

Of course this would involve some testing and confirmation of generated binaries. Given all the other priorities we have, probably pretty low.

Regards,

Willliam

[mentioning @andrewjaykeller @conorrussomanno @produceconsumerobot ]
[package_simblee166_index.json.zip](https://github.com/OpenBCI/OpenBCI_Ganglion_Library/files/2928519/package_simblee166_index.json.zip)

**Comment by shirleyzhang867** [comment]
@wjcroft , 
Thanks for diving into this. I trimmed off the non-Simblee files and[ linked it.](https://drive.google.com/file/d/1FyqoNxvTGLvanM5F23RMR9kC89zmaaKB/view?usp=sharing) The zipped file is now 181 MB. I'll need to look into the suggestions in the latter half of your reply and see about implementing them. Thanks again
looping in @conorrussomanno @produceconsumerobot

**Comment by wjcroft** [comment]
Shirley, thanks.

Since the Simblee.com website was deleted, that removed all their pdf documentation files. I managed to find most of them on the Wayback Machine. Since even there, they are not guaranteed to stay around forever, I've zipped them up here,

https://drive.google.com/file/d/1ZWt6BVOapTPrtRB0AXfj2gOpMzo4fhhi/view?usp=sharing

I also uploaded them to openbci.com/arduino/simblee/ , where we can link to them, for example:

https://openbci.com/arduino/simblee/Simblee_Quickstart_Guide_v1.0.pdf

When and if we customize our own openbci.com/arduino/package_simblee166_index.json to feed the Arduino Board Manager -- we can reference the Simblee_Quickstart_Guide but instruct users to replace the json URL shown there with our own.

The complete list of Simblee files I located, is here:

(71) Simblee - YouTube.webloc
Getting Started with SimbleeCOM v1.0.pdf
Getting Started with SimbleeForMobile v1.02.pdf
News-article-links.webloc  [links off to various internet articles on Simblee development]
Simblee Cloud Reference v1.0.pdf
Simblee Concepts - learn.sparkfun.com.webloc
Simblee EcoSystem Explainer Rev 1.pdf
Simblee Functions Reference V1.0.pdf
Simblee OTA Reference V1.0.pdf
Simblee RFD77101 Datasheet v1.0.pdf
Simblee User Guide v2.05.pdf
Simblee_Quickstart_Guide_v1.0.pdf
Simblee_RFduino_Dual_Footprint_and_Library_Altium_04_14_15.zip
Simblee_RFduino_Dual_Footprint_and_Library_Eagle_04_13_15.zip

Mentioning @conorrussomanno @produceconsumerobot @daniellasry

**Comment by wjcroft** [comment]
This should be a working link for the Arduino IDE Board Manager to build Ganglion,

http://openbci.com/arduino/package_simblee166_index.json

Let us know if this works.

**Comment by wjcroft** [comment]
Does anyone know why there are apparently THREE version numbers associated with the (I assume) last Simblee / Arduino Board Manager package? 

* the Board Manager file is named package_simblee166_index.json
* the version number inside the json is "1.1.4"
* the archive file name is simblee_253.tar.gz

I know that RFDigital folks have scattered to the winds, but still curious. I'm confident we have all the right numbers here in our json link above. 

Mentioning @conorrussomanno @biomurph @produceconsumerobot 

----

Appending my own answer here. The "166" in the json filename is apparently in reference to the Arduino 1.66 IDE which was current when it was created. But we now recommend using the latest 1.8.x IDE. The "1.1.4" is just the sequencing of the Simblee releases. There were earlier ones going back in the original json, but I removed those to avoid confusion. The "253" in the tar.gz file is likely just their other sequencing scheme for the libraries.

**Comment by wjcroft** [comment]
OK! Just installed Simblee files and compiled the Ganglion firmware using the OpenBCI Board Manager JSON. And following the tutorial in,

https://docs.openbci.com/Hardware/09-Ganglion_Programming_Tutorial

Here is the sketch 'Verify' command output:

```
Sketch uses 51844 bytes (39%) of program storage space. Maximum is 131072 bytes.
Global variables use 5624 bytes of dynamic memory.
```

**Let me know if that jives with the Mac output.**

The Ganglion I have here is an old old one, mailed from the Kickstarter that requires some futzing to flash via a serial port instead of OTA. So I stopped at that point.

This was done on a Windows 7 machine with Arduino IDE 1.8.8. In terms of following the tutorial steps for Windows (not Mac as shown in tutorial) -- the variant.h file is located in (with your username):
C:\Users\croft\AppData\Local\Arduino15\packages\Simblee\hardware\Simblee\1.1.4\variants\Simblee

Mentioning @conorrussomanno @produceconsumerobot @shirleyzhang867

**Comment by ronit-malka** [comment]
Hello, the google drive link no longer works (sadly)...for those of us still stuck using a Simblee (may only be me now), is there any way we might have those files elsewhere or could re-upload? Thanks so much in advance!

**Comment by wjcroft** [comment]
@ronit-malka, hi. Please look over this Forum thread below for more tips on Simblee firmware generation. Apologies this is so convoluted. There needs to be an update and consolidation of the build instructions on docs. Especially addressing the variants required per platform: Windows, Mac, Linux. Of course this is all complicated by the fact that Simblee is no longer supported by the chip manufacturer.

https://openbci.com/forum/index.php?p=/discussion/2736/ganglion-ota-programming-for-windows

William

**Comment by ronit-malka** [comment]
Hi William, thanks so much for the reply. I might be missing something massive here, but when I tried downloading http://openbci.com/arduino/simblee_253.tar.gz:
-  copying its contents into the libraries folder (on a Mac: .../Documents/Arduino/libraries) still gives me a compile error, saying I don't have SimbleeCOM.h. 
- copying just the libraries contents into the libraries folder keeps giving compiling errors that it doesn't have access to some of the libraries in the various other folders included in the zip files...when I repeatedly add those into the libraries/SimbleeCOM folder I eventually get an error " #include <sys/types.h>" which is not a file in the folder...

I'm doing this manually because using the openbci simblee package does not seem to be working for adding the board...whether or not I put an http:// or https:// in front, I always get "Error downloading https://openbci.com/arduino/package_simblee166_index.json" (same error, of course, as what I get with the original http://www.simblee.com/package_simblee166_index.json.

In case it's helpful, I'm using Arduino 1.6.7 and MacOSx 10.11.16 - I didn't update *anything* since the last time my Simblee boards worked, and have been hesitant to do so until this project with the boards is over, but updating is the only thing I can think of given this thread and the other thread you provided note the new json might actually be the 1.14 version rather than the 1.6 version?

Thanks again for the help, I really appreciate it!

**Comment by wjcroft** [comment]
> I'm doing this manually because...

Hmm, installing the Simblee board into the Arduino IDE was not meant to be done manually, but with the IDE interpreting the board definition json file. This then pulls in all the related toolchains.

My recommendation at this point is for @shirleyzhang867 to just re-upload or fix the macOS Google Drive file that was originally in place. And is currently referred to in the docs tutorial. I'm sure it is still around, probably just permissions got reset somehow.

The json board file did work for me at one time on Windows. I have not tried recently. I believe it also worked for the user 'teknomage' on that Forum thread I referenced. Though again, he was using Windows, and needed to do some tweaks afterwards to get everything in place.

Regards, William

**Comment by ronit-malka** [comment]
OK, an update: I installed Arduino 1.8.1 and then was able to install the boards perfectly (Arduino —> preferences —> type in http://openbci.com/arduino/package_simblee166_index.json to URL slot and click OK —> Tools —> Board —> Boards Manager —> scroll to bottom and click “install” next to Simblee). Woo! To be clear, this installs the contents of the zip file included above (http://openbci.com/arduino/simblee_253.tar.gz) into ~/Library/Arduino15/packages/Simblee/hardware/Simblee/1.1.4 on my Mac (NOT into the .../Documents/Arduino/libraries folder).

Now I'm able to upload a test ASCII table code to a simblee breakout board but am still unable to compile my old Simblee codes, and am getting the following error: arm-none-eabi-g++: error: core/syscalls.c.o: No such file or directory

This so far hasn't been improved by moving the contents of the core/arduino folder out into the core folder or changing that path in the platform.txt file, but I'm trying to debug now. Using the platform.txt that was modified by teknomage didn't seem to work either, but it was edited for using Gaglion so that might be the issue. If anyone has encountered this before and has suggestions they would be much appreciated!

**Comment by wjcroft** [comment]
@ronit-malka, hi. Shirley fixed the permissions on the Google Drive zip file, https://drive.google.com/file/d/1FyqoNxvTGLvanM5F23RMR9kC89zmaaKB/view

Can you clarify, you mention "...but am still unable to compile my old Simblee codes..." You are referring to Ganglion firmware or something else?

William

**Comment by ronit-malka** [comment]
Thanks Shirley and William! Your help has been invaluable. In terms of what I meant by compiling my old Simblee codes, I had custom PCBs with Simblees on them (totally unrelated to Ganglion, from back when Simblee sold their own boards etc.) and was unable to compile the code for those sketches that were referencing the Simblee compiler.

I actually just got it working without the google drive, in case anyone else has this issue in the future: I ended up copying the compiler section (included below) to END of platform.txt file and it now seems to work just fine. For whatever reason I apparently need to re-save platform.txt every time I reopen Arduino, but it'll now compile sketches designed for Simblee exclusively.

## Combine gc-sections, archives, and objects
recipe.c.combine.pattern="{compiler.path}{compiler.c.elf.cmd}" {compiler.c.elf.flags} -mcpu={build.mcu} {build.extra_flags} "-T{build.variant.path}/{build.ldscript}" "-Wl,-Map,{build.path}/{build.project_name}.map" -Wl,--cref -o "{build.path}/{build.project_name}.elf" "-L{build.path}" -Wl,--warn-common -Wl,--warn-section-align -Wl,--start-group "{build.path}/core/syscalls.c.o" {object_files} "{build.variant.path}/{build.variant_system_lib}" "{build.variant.path}/libSimblee.a" "{build.variant.path}/libSimbleeBLE.a" "{build.variant.path}/libSimbleeGZLL.a" "{build.variant.path}/libSimbleeForMobile.a" "{build.variant.path}/libSimbleeCOM.a" "{build.variant.path}/libSimbleeSensor.a" "{build.path}/{archive_file}" -Wl,--end-group

Thanks again guys!

**Comment by philippitts** [comment]
> OK, an update: I installed Arduino 1.8.1 and then was able to install the boards perfectly (Arduino —> preferences —> type in http://openbci.com/arduino/package_simblee166_index.json to URL slot and click OK —> Tools —> Board —> Boards Manager —> scroll to bottom and click “install” next to Simblee). Woo! To be clear, this installs the contents of the zip file included above (http://openbci.com/arduino/simblee_253.tar.gz) into ~/Library/Arduino15/packages/Simblee/hardware/Simblee/1.1.4 on my Mac (NOT into the .../Documents/Arduino/libraries folder).
> 
> Now I'm able to upload a test ASCII table code to a simblee breakout board but am still unable to compile my old Simblee codes, and am getting the following error: arm-none-eabi-g++: error: core/syscalls.c.o: No such file or directory
> 
> This so far hasn't been improved by moving the contents of the core/arduino folder out into the core folder or changing that path in the platform.txt file, but I'm trying to debug now. Using the platform.txt that was modified by teknomage didn't seem to work either, but it was edited for using Gaglion so that might be the issue. If anyone has encountered this before and has suggestions they would be much appreciated!

If you experience errors downloading the JSON file referenced here, you can try using the below URL instead.

```
https://openbci-simblee.s3.amazonaws.com/package_simblee_index.json
```

This has been confirmed to work on Windows. Will test on MacOS and Ubuntu as well.

**Comment by wjcroft** [comment]
There is a related LONG thread here by teknomage, he had to jump through various hoops to get the Windows build to work. 

https://openbci.com/forum/index.php?p=/discussion/2736/ganglion-ota-programming-for-windows-new-instructions

**Comment by philippitts** [comment]
That's great documentation! I was able to flash on Windows without having to go through all the steps teknomage described since I used the JSON file hosted on AWS. Hopefully this makes it easier for others interested in flashing firmware via Windows. Adding the custom URL and using the board manager provided the Simblee board in Arduino and then after updating the variants.h file as described in the [documentation](https://docs.openbci.com/Ganglion/GanglionProgram/#modify-the-simblee-board-files), I was able to successfully flash (and stream) a Ganglion.

Note that I did this using a hardware jig we have in-house for flashing Ganglion boards. I haven't tried deploying a Windows compiled binary OTA yet.

I'll look at updating the Simblee files we're hosting to have the correct SPI values in the variants.h file. If that works, building custom firmware should be as easy as adding the URL and installing the board.

**Comment by philippitts** [comment]
I updated the package file and edited the Simblee archive so it should work for Ganglion without modification.

To summarize, on any platform follow the steps below:

1. Install Arduino 1.8.x
2. Navigate to File -> Preferences
3. Paste `https://openbci-simblee.s3.amazonaws.com/package_simblee_index.json` into the Additional Boards Manager URLs field
4. Navigate to Tools -> Board -> Boards Manager...
5. Search for Ganglion and install it
6. Select OpenBCI -> Ganglion from the Tools -> Board menu
7. Compile and deploy

Note that to deploy via OTA, you will still need to follow the directions in the [documentation](https://docs.openbci.com/Ganglion/GanglionProgram/#how-to-create-ota-file) (for MacOS) or the [forum thread](https://openbci.com/forum/index.php?p=/discussion/2736/ganglion-ota-programming-for-windows-new-instructions) (for Windows) that @wjcroft mentioned.

We can look into improving the OTA flashing experience in a separate issue.

**Comment by philippitts** [comment]
OTA improvements issue: https://github.com/OpenBCI/OpenBCI_Ganglion_Library/issues/18

**Comment by philippitts** [comment]
Closing this issue as remaining work has been decomposed into separate issues.
